Userland/Libraries/LibSoftGPU/Device.cpp

```diff
@@ -173,6 +173,42 @@ ALWAYS_INLINE static Vector4<f32x4> get_blend_factor(GPU::BlendFactor blend_fact
     }
 }
 
+// OpenGL 2.0 § 4.1.8, table 4.1
+ALWAYS_INLINE static Vector4<f32x4> blend_colors(
+    GPU::BlendEquation rgb_mode,
+    GPU::BlendEquation alpha_mode,
+    Vector4<f32x4> const& source,
+    Vector4<f32x4> const& source_weights,
+    Vector4<f32x4> const& destination,
+    Vector4<f32x4> const& destination_weights)
+{
+    auto apply_equation = [&](GPU::BlendEquation mode, auto source, auto source_weights, auto destination, auto destination_weights) -> auto {
+        switch (mode) {
+        case GPU::BlendEquation::Add:
+            return source * source_weights + destination * destination_weights;
+        case GPU::BlendEquation::Subtract:
+            return source * source_weights - destination * destination_weights;
+        case GPU::BlendEquation::ReverseSubtract:
+            return destination * source_weights - source * destination_weights;
+        case GPU::BlendEquation::Min:
+            return AK::min(source, destination);
+        case GPU::BlendEquation::Max:
+            return AK::max(source, destination);
+        default:
+            VERIFY_NOT_REACHED();
+        }
+    };
+
+    // Single calculation if RGB mode is the same as the alpha mode
+    if (rgb_mode == alpha_mode)
+        return apply_equation(rgb_mode, source, source_weights, destination, destination_weights);
+
+    // Split RGB/alpha calculation
+    auto rgb = apply_equation(rgb_mode, source.xyz(), source_weights.xyz(), destination.xyz(), destination_weights.xyz());
+    auto alpha = apply_equation(alpha_mode, source.w(), source_weights.w(), destination.w(), destination_weights.w());
+    return { rgb.x(), rgb.y(), rgb.z(), alpha };
```

Comment:
Oh, we don't use premul alpha?

Reply:
That's up to you, with `glBlendFunc(Separate)` you can choose how each component is transformed. `glBlendEquation(Separate)` only changes the operation performed on the source color/weights and destination color/weights.